RecoCTPPS/ProtonReconstruction/src/ProtonReconstructionAlgorithm.cc

```diff
@@ -230,66 +256,146 @@ reco::ForwardProton ProtonReconstructionAlgorithm::reconstructFromMultiRP(const
+    fitter_->Config().ParSettings(2).Set("th_y", th_y_init, 2E-6);
+    fitter_->Config().ParSettings(3).Set("vtx_y", vtx_y_init, 10E-6);
+
+    if (!fitVtxY_)
+      fitter_->Config().ParSettings(3).Fix();
+
+    chiSquareCalculator_->tracks = &tracks;
+    chiSquareCalculator_->m_rp_optics = &m_rp_optics_;
+
+    fitter_->FitFCN();
+    fitter_->FitFCN();  // second minimisation in case the first one had troubles
+
+    // extract proton parameters
+    const ROOT::Fit::FitResult &result = fitter_->Result();
+    const double *params = result.GetParams();
+
+    valid = result.IsValid();
+    chi2 = result.Chi2();
+    xi = params[0];
+    th_x = params[1];
+    th_y = params[2];
+    vtx_y = params[3];
+
+    map<unsigned int, signed int> index_map = {
+        {(unsigned int)FP::Index::xi, 0},
+        {(unsigned int)FP::Index::th_x, 1},
+        {(unsigned int)FP::Index::th_y, 2},
+        {(unsigned int)FP::Index::vtx_y, ((fitVtxY_) ? 3 : -1)},
+        {(unsigned int)FP::Index::vtx_x, -1},
+    };
+
+    for (unsigned int i = 0; i < (unsigned int)FP::Index::num_indices; ++i) {
+      signed int fit_i = index_map[i];
+
+      for (unsigned int j = 0; j < (unsigned int)FP::Index::num_indices; ++j) {
+        signed int fit_j = index_map[j];
+
+        cm(i, j) = (fit_i >= 0 && fit_j >= 0) ? result.CovMatrix(fit_i, fit_j) : 0.;
+      }
+    }
+  }
 
-  fitter_->FitFCN();
-  fitter_->FitFCN();  // second minimisation in case the first one had troubles
+  if (multi_rp_algorithm_ == mrpaNewton || multi_rp_algorithm_ == mrpaAnalIter) {
+    // settings
+    const unsigned int maxIterations = 100;
+    const double maxXiDiff = 1E-6;
+    const double xi_ep = 1E-5;
 
-  // extract proton parameters
-  const ROOT::Fit::FitResult &result = fitter_->Result();
-  const double *params = result.GetParams();
+    // collect input
+    double x_N = tracks[0]->x() * 1E-1,  // conversion: mm --> cm
+        x_F = tracks[1]->x() * 1E-1;
+
+    double y_N = tracks[0]->y() * 1E-1,  // conversion: mm --> cm
+        y_F = tracks[1]->y() * 1E-1;
+
+    const RPOpticsData &i_N = m_rp_optics_.find(tracks[0]->rpId())->second,
+                       &i_F = m_rp_optics_.find(tracks[1]->rpId())->second;
+
+    // horizontal reconstruction - run iterations
+    valid = false;
+    double xi_prev = xi_init;
+    for (unsigned int it = 0; it < maxIterations; ++it) {
+      if (multi_rp_algorithm_ == mrpaNewton) {
+        const double g = newtonGoalFcn(xi_prev, x_N, x_F, i_N, i_F);
+        const double gp = (newtonGoalFcn(xi_prev + xi_ep, x_N, x_F, i_N, i_F) - g) / xi_ep;
+
+        xi = xi_prev - g / gp;
+      }
+
+      if (multi_rp_algorithm_ == mrpaAnalIter) {
+        const double D_x_eff_N = i_N.s_x_d_vs_xi->Eval(xi_prev) / xi_prev;
+        const double D_x_eff_F = i_F.s_x_d_vs_xi->Eval(xi_prev) / xi_prev;
+
+        const double L_x_N = i_N.s_L_x_vs_xi->Eval(xi_prev);
+        const double L_x_F = i_F.s_L_x_vs_xi->Eval(xi_prev);
```

Comment:
start local variables and data members with lower case letter. Capitalization is reserved for class names.